### PR TITLE
Support ThemeIcon's `color` property of the VS Code API

### LINKED ISSUE #11128: [vscode] Support optional property `ThemeIcon#color`
### Feature Description:

Support plugins referring to a `ThemeIcon` with a specific `ThemeColor` via constructor or property `color` of a `ThemeIcon`, see https://code.visualstudio.com/api/references/vscode-api#ThemeIcon

### PR DESCRIPTION
**Pull request by lucas-koehler**
#### What it does
Fixes #11128

* Add optional `color` property to `ThemeIcon` of the VS Code API
* Hand over the `ThemeIcon` instead of just its id from plugin Ext to Main
* Apply the color in the `PluginTree`
* Adapt `TreeViewNode.is` to also recognize sub types such as `CompositeTreeViewNode`

Contributed on behalf of STMicroelectronics

Signed-off-by: Lucas Koehler <lkoehler@eclipsesource.com>

#### How to test

* Download the [test extension](https://github.com/lucas-koehler/vscode-extension-samples/releases/download/theme-icon-color-test-0.0.1/custom-view-samples-0.0.1.vsix). [Source](https://github.com/lucas-koehler/vscode-extension-samples/tree/theme-icon-color-test/tree-view-sample) for reference.
* Launch and open the `TEST VIEW' in the Explorer:
image

* Expand the tree nodes. See that the tree nodes use file icons in the editor's warning color (usually some kind of yellow or orange)
* See that file icons in other views (e.g. the File Explorer) still have their original color
* Switch themes to see that the icons in the Test View use the warning color of the selected theme (select high contrast theme to see a noticeable difference)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin/src/theia.d.ts [hunk 1]
```diff
@@ -2652,7 +2652,7 @@ export module '@theia/plugin' {
          */
         static readonly Folder: ThemeIcon;
 
-        private constructor(public id: string);
+        private constructor(public id: string, public color?: ThemeColor);
```
Comment (vince-fugnitto) [inline packages/plugin/src/theia.d.ts:2655]: minor: do you mind adding the documentation here?

```suggestion
		/**
		 * Creates a reference to a theme icon.
		 * @param id id of the icon. The available icons are listed in https://code.visualstudio.com/api/references/icons-in-labels#icon-listing.
		 * @param color optional `ThemeColor` for the icon. The color is currently only used in {@link TreeItem}.
		 */
        private constructor(public id: string, public color?: ThemeColor);
```

Comment (vince-fugnitto) [inline packages/plugin/src/theia.d.ts:2655]: I believe you are missing the optional `color` in `ThemeIcon` (which will still report as missing in the compatibility report):

```ts
/**
 * The optional ThemeColor of the icon. The color is currently only used in {@link TreeItem}.
 */
readonly color?: ThemeColor | undefined;
```

Comment (lucas-koehler) [reply packages/plugin/src/theia.d.ts:2655]: Are you sure we need this? There is no explicit property for `id` either and it is recognized in the report. Or is this different for optional properties? In this case I'll add it. I would then also add the `id` prop to make it consistent with the VS Code API which defines both explicitly.

File: packages/plugin-ext/src/main/browser/view/plugin-tree-view-node-label-provider.ts [hunk 2]
```diff
@@ -43,12 +43,14 @@ export class PluginTreeViewNodeLabelProvider implements LabelProviderContributio
         if (node.icon) {
             return node.icon;
         }
-        if (node.themeIconId) {
-            if (node.themeIconId === 'file' || node.themeIconId === 'folder') {
+        if (node.themeIcon) {
+            if (node.themeIcon.id === 'file' || node.themeIcon.id === 'folder') {
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-tree-view-node-label-provider.ts:47]: I just wanted to highlight that this only allows us a subset of icons to be used which is different than the API actually allows. VS Code allows all icons listed here (https://code.visualstudio.com/api/references/icons-in-labels#icon-listing) which is documented in `id`:

```ts
/**
 * The id of the icon. The available icons are listed in https://code.visualstudio.com/api/references/icons-in-labels#icon-listing.
 */
readonly id: string;
```

With some modifications to the extension provided in VS Code I can see (while in Theia it does not work):


image

**Review by vince-fugnitto — COMMENTED** [review]
I confirm that the `color` is working well with the sample extension.

I made a comment about actually properly supporting all options for icons however that we can possibly tackle later if you don't want to update.

**Comment by lucas-koehler** [comment]
@vince-fugnitto
>I made a comment about actually properly supporting all options for icons however that we can possibly tackle later if you don't want to update.

I would prefer to have that as a follow up because it does not seem trivial to support all these icons. For instance, I investigated a bit with the example you provided. The `accounts-view-bar-icon` theme icon is contributed in VSCode as part of its `ActivitybarPart`. It basically maps this identifier to the account codicon. I guess this is not even known to Theia (and possibly many other icons made available in such a way in the VS Code workbench).

**Review by vince-fugnitto — APPROVED** [review]
The changes look good to me, I confirmed that the color is properly applied :+1: 
I did not notice any regressions.

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx [hunk 3]
```diff
@@ -60,15 +61,15 @@ export interface TreeViewNode extends SelectableTreeNode {
...
-    themeIconId?: string | 'folder' | 'file';
+    themeIcon?: ThemeIcon;
     tooltip?: string;
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     description?: string | boolean | any;
     accessibilityInformation?: AccessibilityInformation;
 }
 export namespace TreeViewNode {
     export function is(arg: TreeNode | undefined): arg is TreeViewNode {
-        return !!arg && SelectableTreeNode.is(arg) && !ExpandableTreeNode.is(arg) && !CompositeTreeNode.is(arg);
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx:71]: @lucas-koehler can you clarify why this change is necessary?
